public-images/public-images-explanation/lxd-openstack-images.rst

```diff
@@ -0,0 +1,210 @@
+
+1. Uncompress the LXD tarball.
+2. Make modifications to the ``metadata.yaml`` file. See the `LXD image
+   format
+   documentation <https://documentation.ubuntu.com/lxd/en/latest/reference/image_format/#>`__
+   to learn more about image metadata and the templates you may wish to
+   modify.
+3. Compress the metadata and templates.
+
+This snippet from the `How to customise LXD image metadata for
+cloud-init <https://discourse.ubuntu.com/t/how-to-customize-lxd-image-metadata-for-cloud-init/25157>`__
+guide referenced below demonstrates a typical workflow:
+
+.. code :: bash
+
+   # Uncompress original LXD metadata
+   $ tar xf ${RELEASE}-server-cloudimg-amd64-lxd.tar.xz
+   # Add directives to create /etc/cloud/cloud.cfg.d/95-use-lxd.cfg
+   $ cat > templates/cloud-init-use-lxd.tpl <<EOF
+   # Added by LXD metadata.yaml
+   datasource_list: [ LXD, NoCloud ]
+   EOF
+   $ cat > add-lxd.yaml <<EOF
+       /etc/cloud/cloud.cfg.d/95-use-lxd.cfg:
+           when:
+               - create
+               - copy
+           template: cloud-init-use-lxd.tpl
+   EOF
+   $ cat add-lxd.yaml >> metadata.yaml
+   # Compress LXD metadata and templates
+   $ tar -czf ${RELEASE}-server-cloudimg-amd64-prefer-lxd.tar.xz metadata.yaml templates/
+
+
+Cloud-init
+^^^^^^^^^^
+
+`Cloud-init <https://cloudinit.readthedocs.io/en/latest/index.html>`__
+is used to initialise cloud instances on first boot. A detailed guide on
+customising LXD image metadata for ``cloud-init`` can be found
+`here <https://discourse.ubuntu.com/t/how-to-customize-lxd-image-metadata-for-cloud-init/25157>`__.
+If you want to configure ``cloud-init`` once an instance has been
+created (but not booted), refer to the LXD docs
+`here <https://documentation.ubuntu.com/lxd/en/latest/cloud-init/>`__.
+
+
+CLI
+^^^
+
+The `LXD documentation on
+images <https://documentation.ubuntu.com/lxd/en/latest/howto/images_manage/>`__
+has an extensive guide on managing images. Two of the most common use
+cases are to set individual properties or to edit all of the image
+properties.
+
+.. code :: bash
+
+   # set a specific image property
+   lxc image set-property <image_ID> <key> <value>
+
+
+.. code :: bash
+
+   # edit the full image properties
+   lxc image edit <image_ID>
+
+
+OpenStack images
+~~~~~~~~~~~~~~~~
+
+OpenStack has an extensive guide on `modifying
+images <https://docs.openstack.org/image-guide/modify-images.html>`__
+that is applicable to the QCOW images Ubuntu provides. It leverages ``libguestfs`` 
+`tools <https://libguestfs.org/>`__ in order to access and
+modify disk images. You can use the ``guestfish``
+`interactive shell <https://libguestfs.org/guestfish.1.html>`__ 
+— which exposes the full functionality of the ``guestfs`` API — or
+rely on the ``virt-*`` tools from ``libguestfs`` — like ``virt-cat`` for
+displaying files, ``virt-df`` for checking free space, and
+``virt-inspector`` for inspecting VM images — to perform specific tasks.
```

Comment:
```suggestion
``virt-inspector`` for inspecting VM images.
```

Reply:
Whoops, many changes to this sentence and this one slipped.

I usually fix in one commit and then rebase to keep the commit history clean, and haven't looked into doing that via the GitHub UI.

Reply:
Using the `Add suggestion to batch` in the UI option allows you to create a single commit. But yeah you'll still need to squash commits to keep the commit history clean and run the spell/link checks locally.